lib/gems/pending/util/object_storage/miq_swift_storage.rb

```diff
@@ -0,0 +1,148 @@
+    @password       = @settings[:password]
+    @container_name = @mount_path[0] == File::Separator ? @mount_path[1..-1] : @mount_path
+  end
+
+  def uri_to_local_path(remote_file)
+    # Strip off the leading "swift:/" from the URI"
+    File.join(@mnt_point, URI(remote_file).host, URI(remote_file).path)
+  end
+
+  def uri_to_object_path(remote_file)
+    # Strip off the leading "swift://" and the container name from the URI"
+    # Also remove teh leading delimiter.
+    object_file_with_bucket = URI.split(URI.encode(remote_file))[5]
+    object_file_with_bucket.split(File::Separator)[2..-1].join(File::Separator)
+  end
+
+  def upload_single(dest_uri)
+    #
+    # Get the remote path, and parse out the bucket name.
+    #
+    object_file = uri_to_object_path(dest_uri)
+    #
+    # write dump file to swift
+    #
+    logger.debug("Writing [#{source_input}] to => Bucket [#{container_name}] using object file name [#{object_file}]")
+    begin
+      swift_file = container.files.new(:key => object_file)
+      params     = {
+        :expects       => [201, 202],
+        :headers       => {},
+        :request_block => -> { read_single_chunk },
+        :idempotent    => false,
+        :method        => "PUT",
+        :path          => "#{Fog::OpenStack.escape(swift_file.directory.key)}/#{Fog::OpenStack.escape(swift_file.key)}"
+      }
+      swift_file.service.send(:request, params)
+      clear_split_vars
+    rescue Excon::Errors::Unauthorized => err
+      logger.error("Access to Swift container #{@container_name} failed due to a bad username or password. #{err}")
+      msg = "Access to Swift container #{@container_name} failed due to a bad username or password. #{err}"
+      raise err, msg, err.backtrace
+    rescue => err
+      logger.error("Error uploading #{source_input} to Swift container #{@container_name}. #{err}")
+      msg = "Error uploading #{source_input} to Swift container #{@container_name}. #{err}"
+      raise err, msg, err.backtrace
+    end
+  end
+
+  def mkdir(_dir)
+    container
+  end
+
+  def container
+    @container ||= begin
+                     container   = swift.directories.get(container_name)
+                     logger.debug("Swift container [#{container}] found") if container
+                     container ||= create_container
+                     container
+                   rescue Fog::Storage::OpenStack::NotFound
+                     logger.debug("Swift container #{container_name} does not exist.  Creating.")
+                     create_container
+                   rescue => err
+                     logger.error("Error getting Swift container #{container_name}. #{err}")
+                     msg = "Error getting Swift container #{container_name}. #{err}"
+                     raise err, msg, err.backtrace
+                   end
+  end
+
+  private
+
+  def swift
+    return @swift if @swift
+    require 'manageiq/providers/openstack/legacy/openstack_handle'
```

Comment:
So of note, I have written a version of this that makes use of `fog/openstack` directly:

https://github.com/ManageIQ/manageiq-gems-pending/blob/cd2beb72ae0e0c8cc96c39362165350371d77501/lib/gems/pending/util/object_storage/miq_swift_storage.rb#L92-L114

And seems to work with the tests I have setup for this (rake tasks only).  If we want to limit the amount of dependencies from that repo, that might be something to consider.  I spent a decent amount of time pulling out the relevant pieces from the `OpenstackHandle` code base to make that work, so hopefully that does the trick for most use cases.

Reply:
:+1: For @NickLaMuro's approach. Based on the snippet linked here, I don't see much simplification in the way the code is currently vs using `fog/openstack`.